File: src/cardinal.ts
import { NumberToWordsOptions, defaultOptions } from "./options";

const ones = [
  "",
  "one",
  "two",
  "three",
  "four",
  "five",
  "six",
  "seven",
  "eight",
  "nine",
  "ten",
  "eleven",
  "twelve",
  "thirteen",
  "fourteen",
  "fifteen",
  "sixteen",
  "seventeen",
  "eighteen",
  "nineteen",
];

const tens = [
  "",
  "",
  "twenty",
  "thirty",
  "forty",
  "fifty",
  "sixty",
  "seventy",
  "eighty",
  "ninety",
];

const scales = [
  "",
  "thousand",
  "million",
  "billion",
  "trillion",
  "quadrillion",
  "quintillion",
  "sextillion",
  "septillion",
  "octillion",
  "nonillion",
  "decillion",
];

function convertLessThanOneThousand(
  number: number,
  options: NumberToWordsOptions = {},
): string {
  const mergedOptions = { ...defaultOptions, ...options };
  if (number === 0) return "";
  let words = "";
  if (number < 20) {
    words = ones[number];
  } else if (number < 100) {
    const remainder = number % 10;
    words = tens[Math.floor(number / 10)];
    if (remainder > 0) {
      words += "-" + ones[remainder];
    }
  } else {
    const remainder = number % 100;
    words = ones[Math.floor(number / 100)] + " hundred";
    if (remainder > 0) {
      words +=
        (mergedOptions.useAnd ? " and " : " ") +
        convertLessThanOneThousand(remainder, mergedOptions);
    }
  }
  return words;
}

export function toWords(
  number: number | string,
  options: NumberToWordsOptions = {},
): string {
  let num: number;
  if (typeof number === "string") {
    num = parseInt(number, 10);
    if (isNaN(num)) throw new Error("Invalid number string provided");
  } else if (typeof number === "number") {
    if (!isFinite(number) || isNaN(number))
      throw new Error("Invalid number provided");
    num = number;
  } else {
    throw new Error(
      "Input must be a number or a string representation of a number",
    );
  }
  num = Math.floor(num);
  const mergedOptions = { ...defaultOptions, ...options };
  if (num === 0) return "zero";
  let prefix = "";
  if (num < 0) {
    prefix = "negative ";
    num = Math.abs(num);
  }
  let words = "";
  let scaleIndex = 0;
  while (num > 0) {
    const chunk = num % 1000;
    if (chunk > 0) {
      const chunkWords = convertLessThanOneThousand(chunk, mergedOptions);
      const scale = scales[scaleIndex];
      words =
        chunkWords +
        (scale ? " " + scale : "") +
        (words && mergedOptions.useCommas ? ", " : words ? " " : "") +
        words;
    }
    scaleIndex++;
    num = Math.floor(num / 1000);
  }
  let result = prefix + words.trim();
  if (mergedOptions.appendOnly) result += " only";
  if (mergedOptions.capitalize)
    result = result.charAt(0).toUpperCase() + result.slice(1);
  if (mergedOptions.uppercase) result = result.toUpperCase();
  return result;
}

export { scales, ones, tens };

File: packages/numlang-rs/src/cardinal.rs
use crate::Options;

pub const ONES: [&str; 20] = [
    "",
    "one",
    "two",
    "three",
    "four",
    "five",
    "six",
    "seven",
    "eight",
    "nine",
    "ten",
    "eleven",
    "twelve",
    "thirteen",
    "fourteen",
    "fifteen",
    "sixteen",
    "seventeen",
    "eighteen",
    "nineteen",
];

pub const TENS: [&str; 10] = [
    "", "", "twenty", "thirty", "forty", "fifty", "sixty", "seventy", "eighty", "ninety",
];

pub const SCALES: [&str; 12] = [
    "",
    "thousand",
    "million",
    "billion",
    "trillion",
    "quadrillion",
    "quintillion",
    "sextillion",
    "septillion",
    "octillion",
    "nonillion",
    "decillion",
];

fn convert_less_than_one_thousand(n: u16, options: &Options) -> String {
    if n == 0 {
        return "".to_string();
    }
    if n < 20 {
        return ONES[n as usize].to_string();
    }
    if n < 100 {
        let ten = n / 10;
        let rem = n % 10;
        if rem == 0 {
            TENS[ten as usize].to_string()
        } else {
            format!("{}-{}", TENS[ten as usize], ONES[rem as usize])
        }
    } else {
        let hundred = n / 100;
        let rem = n % 100;
        let and = if options.use_and && rem > 0 {
            " and "
        } else {
            " "
        };
        if rem == 0 {
            format!("{} hundred", ONES[hundred as usize])
        } else {
            format!(
                "{} hundred{}{}",
                ONES[hundred as usize],
                and,
                convert_less_than_one_thousand(rem, options)
            )
        }
    }
}

fn convert_decimal_part(decimal: &str) -> String {
    decimal
        .chars()
        .filter_map(|c| c.to_digit(10))
        .map(|d| if d == 0 { "zero" } else { ONES[d as usize] })
        .collect::<Vec<_>>()
        .join(" ")
}

pub fn to_words(number: impl Into<f64>, options: &Options) -> String {
    let num = number.into();
    let mut prefix = String::new();
    let mut abs_num = num;
    if num < 0.0 {
        prefix = "negative ".to_string();
        abs_num = -num;
    }

    // Convert to string to preserve decimal precision
    let num_str = format!("{}", abs_num);
    let parts: Vec<&str> = num_str.split('.').collect();
    let int_part = parts[0];
    let decimal_part = if parts.len() > 1 { parts[1] } else { "" };

    let int_num: i64 = int_part.parse().unwrap_or(0);

    let mut words = if int_num == 0 {
        "zero".to_string()
    } else {
        let mut n = int_num;
        let mut scale = 0;
        let mut chunks = Vec::new();
        while n > 0 {
            let chunk = (n % 1000) as u16;
            if chunk > 0 {
                let mut chunk_str = convert_less_than_one_thousand(chunk, options);
                if !SCALES[scale].is_empty() {
                    chunk_str.push(' ');
                    chunk_str.push_str(SCALES[scale]);
                }
                chunks.push(chunk_str);
            }
            n /= 1000;
            scale += 1;
        }
        chunks.reverse();
        if options.use_commas {
            chunks.join(", ")
        } else {
            chunks.join(" ")
        }
    };

    if !decimal_part.is_empty() {
        words.push_str(" point ");
        words.push_str(&convert_decimal_part(decimal_part));
    }

    let mut result = prefix + &words;
    if options.append_only {
        result.push_str(" only");
    }
    if options.capitalize && !result.is_empty() {
        let mut chars = result.chars();
        if let Some(first) = chars.next() {
            result = first.to_uppercase().collect::<String>() + chars.as_str();
        }
    }
    if options.uppercase {
        result = result.to_uppercase();
    }
    result
}
